src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,453 @@
+    if (!playerData->GetTotalReports())
+        playerData->SetCreationTime(getMSTime());
+
+    // increasing total_reports
+    playerData->SetTotalReports(playerData->GetTotalReports() + 1);
+    // increasing specific cheat report
+    playerData->SetTypeReports(type, playerData->GetTypeReports(type) + 1);
+
+    // diff time for average calculation
+    uint32 diffTime = getMSTimeDiff(playerData->GetCreationTime(), getMSTime()) / IN_MILLISECONDS;
+
+    if (diffTime > 0)
+    {
+        // Average == Reports per second
+        float average = float(playerData->GetTotalReports()) / float(diffTime);
+        playerData->SetAverage(average);
+    }
+
+    if (sWorld->getIntConfig(CONFIG_ANTICHEAT_MAX_REPORTS_FOR_DAILY_REPORT) < playerData->GetTotalReports())
+    {
+        if (!playerData->GetDailyReportState())
+        {
+            SQLTransaction trans = CharacterDatabase.BeginTransaction();
+            PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_AC_PLAYER_DAILY_REPORTS);
+            stmt->setUInt32(0, player->GetGUID().GetCounter());
+            trans->Append(stmt);
+
+            uint8 index = 0;
+            stmt = CharacterDatabase.GetPreparedStatement(CHAR_INS_AC_PLAYER_DAILY_REPORTS);
+            stmt->setUInt32(index,   player->GetGUID().GetCounter());
+            stmt->setFloat (++index, playerData->GetAverage());
+            stmt->setUInt32(++index, playerData->GetTotalReports());
+            for (uint8 i = 0; i < MAX_REPORT_TYPES; ++i)
+                stmt->setUInt32(++index, playerData->GetTypeReports(static_cast<ReportTypes>(i)));
+
+            stmt->setUInt32(++index, playerData->GetCreationTime());
+            trans->Append(stmt);
+
+            CharacterDatabase.CommitTransaction(trans);
+            playerData->SetDailyReportState(true);
+        }
+    }
+
+    if (sendReport && playerData->GetTotalReports() > sWorld->getIntConfig(CONFIG_ANTICHEAT_REPORTS_INGAME_NOTIFICATION))
+    {
+        playerData->SetLastReportTimer(getMSTime());
+
+        WorldPacket data;
+        std::string str = Trinity::StringFormat(sObjectMgr->GetTrinityStringForDBCLocale(LANG_AC_CHEAT_REPORT), player->GetName().c_str(), player->GetName().c_str());
+        ChatHandler::BuildChatPacket(data, CHAT_MSG_SYSTEM, LANG_UNIVERSAL, NULL, NULL, str);
+        sWorld->SendGlobalGMMessage(&data);
+    }
+}
+
+bool SpeedHackCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    AnticheatData* playerData = GetPlayerData(player->GetGUID());
+    if (!playerData)
+        return false;
+
+    // We also must check the map because the movementFlag can be modified by the client.
+    // If we just check the flag, they could always add that flag and always skip the speed hacking detection.
+    // 369 == DEEPRUN TRAM
+    if (player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_ONTRANSPORT) && player->GetMapId() == 369)
```

Comment:
Yes but only MO_TRANSPORT are Transport* - this kind of check/exclusion is useless if you cant verify if the player really is on the tram or not (the cheater would just need to send  ONTRANSPORT flag with any random transport gameobject guid present on map and he would be able to speedhack all he wants)